crates/circuit/src/dag_circuit.rs

```diff
@@ -2631,6 +2632,96 @@ def _format(operand):
                         | [OperationRef::Instruction(_), OperationRef::StandardInstruction(_)] => {
                             Ok(inst1.py_op_eq(py, inst2)? && check_args() && check_conditions()?)
                         }
+                        [OperationRef::Unitary(op_a), OperationRef::Unitary(op_b)] => {
+                            match [&op_a.array, &op_b.array] {
+                                [ArrayType::NDArray(a), ArrayType::NDArray(b)] => {
+                                    Ok(relative_eq!(a, b, max_relative = 1e-5, epsilon = 1e-8))
+                                }
+                                [ArrayType::NDArray(a), ArrayType::OneQ(b)] => {
+                                    if a.shape()[0] == 2 {
+                                        for i in 0..2 {
+                                            for j in 0..2 {
+                                                if !relative_eq!(
+                                                    a[[i, j]],
+                                                    b[(i, j)],
+                                                    max_relative = 1e-5,
+                                                    epsilon = 1e-8
+                                                ) {
+                                                    return Ok(false);
+                                                }
+                                            }
+                                        }
+                                        Ok(true)
+                                    } else {
+                                        Ok(false)
+                                    }
+                                }
+                                [ArrayType::NDArray(a), ArrayType::TwoQ(b)] => {
+                                    if a.shape()[0] == 4 {
+                                        for i in 0..4 {
+                                            for j in 0..4 {
+                                                if !relative_eq!(
+                                                    a[[i, j]],
+                                                    b[(i, j)],
+                                                    max_relative = 1e-5,
+                                                    epsilon = 1e-8
+                                                ) {
+                                                    return Ok(false);
+                                                }
+                                            }
+                                        }
+                                        Ok(true)
+                                    } else {
+                                        Ok(false)
+                                    }
+                                }
+                                [ArrayType::OneQ(a), ArrayType::NDArray(b)] => {
```

Comment:
Can you combine arms like this to avoid duplication? I think Rust should be happy with this since the types will have the same types for both matches

```rust
[ArrayType::OneQ(one_q), ArrayType::NDArray(nd)] | [ArrayType::NDArray(nd), ArrayType::OneQ(one_q)]
```

Reply:
I wasn't sure if the compiler was smart enough to resolve the typing in cases like this. But I can give a try, I'll update it if it's happy and report back if it's not.

Reply:
Yeah it works, done in: https://github.com/Qiskit/qiskit/pull/13759/commits/e1d6b9b2d3c0bd2f13bd7bb00836056092552d70